client/src/components/examples/MatchingRequestTableContainer.tsx

```diff
@@ -0,0 +1,236 @@
+            _id: "60e3b9596cd5ae8396bd3f53"
+        },
+        createdAt: new Date(1579697012390),
+        deleted: false,
+        fulfilled: false,
+        matchedDonations: [{ donationForm: "60fcbbef3e3ece32ac7be357", quantity: 3 }]
+    },
+    {
+        _id: "60e3b95a6cd5ae8396bd3f5b",
+        quantity: 10,
+        clientName: "Wilma Boyer",
+        requestType: {
+            _id: "60e3b9596cd5ae8396bd3f53"
+        },
+        createdAt: new Date(1608736758629),
+        deleted: false,
+        fulfilled: false,
+        matchedDonations: [{ donationForm: "60fcbbee3e3ece32ac7be355", quantity: 3 }]
+    },
+    {
+        _id: "60e3b95a6cd5ae8396bd3f5d",
+        quantity: 3,
+        clientName: "Mariane Beer",
+        requestType: {
+            _id: "60e3b9596cd5ae8396bd3f53"
+        },
+        createdAt: new Date(1566586665253),
+        deleted: false,
+        fulfilled: false,
+        matchedDonations: [{ donationForm: "60fcbbef3e3ece32ac7be357", quantity: 1 }]
+    }
+];
+
+const MatchingRequestTableContainer: FunctionComponent<Record<string, never>> = () => {
+    const [requests, setRequests] = useState(sampleRequests);
+    const [donationForms, setDonationForms] = useState(sampleDonationForms);
+    const [curDonationForm, setCurDonationForm] = useState(donationForms[1]);
+
+    useEffect(() => {
+        // just here to prevent eslint warnings
+        setDonationForms(donationForms);
+        setCurDonationForm(donationForms[1]);
+    }, []);
+
+    const onQuantitySelected = (quantity: number, requestId: string) => {
+        // find the index of the updated request
+        const reqIndex = requests.findIndex((req) => req._id == requestId);
+        const req = requests[reqIndex];
+
+        const contributionIndex = req?.matchedDonations?.findIndex(
+            (contrib) => contrib.donationForm == curDonationForm._id
+        ) as number;
+
+        // update the matching selection
+        const newMatches = req?.matchedDonations;
+        if (newMatches) {
+            if (contributionIndex != -1) {
+                newMatches[contributionIndex].quantity = quantity;
+            } else {
+                newMatches.push({ donationForm: curDonationForm._id as string, quantity });
+            }
+        }
+
+        // update requests with new donation matches
+        setRequests([
+            ...requests.slice(0, reqIndex),
+            {
+                ...req,
+                matchedDonations: newMatches
+            },
+            ...requests.slice(reqIndex + 1)
+        ]);
+    };
+
+    return (
+        <div
+            style={{
+                marginTop: "30px",
+                marginLeft: "30px",
+                display: "flex",
+                flexDirection: "column",
+                width: "60%"
+            }}
+        >
+            <h3>Not currently matching</h3>
+            <DonationMatchingRequestsTable requests={requests} onQuantitySelected={() => {}} isErroneous={false} />
```

Comment:
Just a comment:

The container of the DonationMatchingRequestsTable should handle the logic for GENERAL (req group is null) and SPECIFIC (req group is not null - where request types matters) donation forms.

Reply:
That's a good point! I'm just using this component as an example to display the table, so it doesn't reflect the actual component hierarchy. For the actual implementation I'm thinking of an approach like this:

`AdminDonationMatchingPage.tsx`
- page component that handles the logic for fetching donation forms and requests, as well as logic for error detection and saving the matches to the backend
- essentially a wrapper for determining what "state" the matching page is in, which is one of the following:
    1. User is matching a specific donation form (donationForm not null, req group not null)
    2. User is matching a general donation form (donationForm not null, req group is null)
    3. User is viewing all available donation forms of a specific request group e.g. all Antique robes (donationForm null, req group not null)

The subcomponents will render different views based on the above states:

Left side: `MatchingRequestsView.tsx`
- determines which view to display for the left side of the page
- If requestGroup is not null (state i or iii), then will render a list of dropdowns based on the requestType
- otherwise (state ii), render a single table with all the requests

Right side: `MatchingDonationFormView.tsx`
- if donationForm is not null (state i or ii), display the information for the single donation form (since the user is currently matching)
- otherwise (state iii), display a list of available donations

sorry for the big wall of text, just wanted to clearly explain what I was thinking. lmk if any part doesn't make sense!

Reply:
This approach sounds great! Appreciate the detail as well :)